nucliadb_node/src/merge/scheduler.rs

```diff
@@ -0,0 +1,294 @@
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+//
+
+use std::fs;
+use std::sync::Condvar;
+use std::time::Duration;
+use std::{
+    path::PathBuf,
+    sync::{Arc, Mutex},
+};
+
+use anyhow::anyhow;
+use nucliadb_core::vectors::MergeMetrics;
+use tokio::sync::oneshot;
+use tokio::sync::oneshot::{Receiver, Sender};
+
+use crate::shards::ShardId;
+use crate::{settings::Settings, shards::providers::shard_cache::ShardWriterCache};
+use nucliadb_core::tracing::warn;
+use nucliadb_core::NodeResult;
+
+use super::work::WorkQueue;
+use crate::merge::{MergePriority, MergeRequest, MergeWaiter};
+
+// Time between scheduler being idle and scheduling all shards for merge
+#[cfg(not(test))]
+const SCHEDULE_ALL_SHARDS_DELAY: Duration = Duration::from_secs(10);
+#[cfg(test)]
+const SCHEDULE_ALL_SHARDS_DELAY: Duration = Duration::from_millis(50);
+
+/// Merge scheduler is the responsible for scheduling merges in the vectors
+/// index. When running, it takes the most prioritary merge request, takes a
+/// shard writer and executes the merge.
+pub struct MergeScheduler {
+    work_queue: Mutex<WorkQueue<InternalMergeRequest>>,
+    shard_cache: Arc<ShardWriterCache>,
+    settings: Settings,
+    condvar: Condvar,
+    shutdown: Mutex<bool>,
+}
+
+impl MergeScheduler {
+    pub fn new(shard_cache: Arc<ShardWriterCache>, settings: Settings) -> Self {
+        Self {
+            work_queue: Mutex::new(WorkQueue::new()),
+            shard_cache,
+            settings,
+            condvar: Condvar::default(),
+            shutdown: Mutex::new(false),
+        }
+    }
+
+    /// Schedule a new merge request. The work handle returned can be used to
+    /// wait for the merge result.
+    pub fn schedule(&self, request: MergeRequest) -> WorkHandle {
+        let (request, priority, handle) = self.prepare(request);
+
+        let mut queue = self.work_queue.lock().expect("Poisoned merger scheduler mutex");
+        queue.push(request, priority);
+        drop(queue);
+
+        // Awake scheduler if it's asleep
+        self.condvar.notify_all();
+
+        handle
+    }
+
+    /// Runs the scheduler until stopped
+    pub fn run(&self) {
+        let mut shutdown = self.shutdown.lock().expect("Poisoned merger shutdown signal");
+        *shutdown = false;
+        drop(shutdown);
+
+        loop {
+            let merge = self.process();
+            if let Err(error) = merge {
```

Comment:
`shutdown` should be read somewhere around here to exit the loop if it's set to true.